astropy/utils/parsing.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Wrappers for PLY to provide thread safety.
+"""
+
+import contextlib
+import functools
+import re
+import os
+import threading
+
```

Comment:
Need an `__all__` to guard against star import by user, and also to control what is visible in API doc. I added the public API in here, but please adjust as you see fit.
```suggestion

__all__ = ['lex', 'ThreadSafeParser', 'yacc']
```

Reply:
Done (also added TAB_HEADER since @mhvk requested that be public, although there is still some discussion ongoing).

Reply:
Given the logic above, let's keep `_TAB_HEADER` out of this (with an `__all__`, we can in principle remove the leading underscore, though it doesn't matter much).